compiler/noirc_frontend/src/elaborator/enums.rs

```diff
@@ -722,16 +794,38 @@ impl Elaborator<'_> {
     /// This is an adaptation of https://github.com/yorickpeterse/pattern-matching-in-rust/tree/main/jacobs2021
     /// which is an implementation of https://julesjacobs.com/notes/patternmatching/patternmatching.pdf
     pub(super) fn elaborate_match_rows(&mut self, rows: Vec<Row>) -> HirMatch {
-        self.compile_rows(rows).unwrap_or_else(|error| {
-            self.push_err(error);
-            HirMatch::Failure
-        })
+        MatchCompiler::run(self, rows)
+    }
+}
+
+impl<'elab, 'ctx> MatchCompiler<'elab, 'ctx> {
+    fn run(elaborator: &'elab mut Elaborator<'ctx>, rows: Vec<Row>) -> HirMatch {
+        let mut compiler = Self {
+            elaborator,
+            has_missing_cases: false,
+            unreachable_cases: rows.iter().map(|row| (row.body, row.location)).collect(),
+        };
+
+        let hir_match = compiler.compile_rows(rows).unwrap_or_else(|error| {
+            compiler.elaborator.push_err(error);
+            HirMatch::Failure { missing_case: false }
+        });
+
+        if compiler.has_missing_cases {
```

Comment:
```suggestion
        if let HirMatch::Failure { missing_case: true } = hir_match {
            compiler.issue_missing_cases_error(&hir_match);
        }
```
nit: This is the only place I see `has_missing_cases` is read. We could remove this field entirely by just check the match status.

Reply:
It's possible the missing case is nested within `hir_match` though. This is the case when you have a match expression like `match x { Some(None) => (), None => (), }` and we're missing a `Some(Some(_))` case